e2e-chatbot-app/app.py

```diff
@@ -0,0 +1,182 @@
+    for delta in deltas[1:]:
+        if delta.tool_calls:
+            result_msg = result_msg.copy(update={"tool_calls": delta.tool_calls})
+        if delta.tool_call_id:
+            result_msg = result_msg.copy(update={"tool_call_id": delta.tool_call_id})
+        msg_contents.append(delta.content)
+    result_msg = result_msg.copy(update={"content": "".join(msg_contents)})
+    return result_msg
+
+
+class UserMessage:
+    def __init__(self, content):
+        self.content = content
+
+    def to_input_messages(self):
+        return [{
+            "role": "user",
+            "content": self.content
+        }]
+
+    def render(self, idx):
+        with st.chat_message("user"):
+            st.markdown(self.content)
+
+def render_message(msg):
+    if msg["role"] == "assistant" and "tool_calls" in msg:
+        for call in msg["tool_calls"]:
+            fn_name = call["function"]["name"]
+            args = call["function"]["arguments"]
+            st.markdown(msg["content"])
+            st.markdown(f"🛠️ Calling **`{fn_name}`** with:\n```json\n{args}\n```")
+    elif msg["role"] == "tool":
+        st.markdown("🧰 Tool Response:")
+        st.code(msg["content"], language="json")
+    elif msg["role"] == "assistant" and msg.get("content"):
+        st.markdown(msg["content"])
+
+class AssistantResponse:
+    def __init__(self, messages, request_id):
+        self.messages = messages
+        self.request_id = request_id
+
+    def to_input_messages(self):
+        return self.messages
+
+    def render(self, idx):
+        with st.chat_message("assistant"):
+            for msg in self.messages:
+                render_message(msg)
+
+            if self.request_id is not None:
+                render_assistant_message_feedback(idx, self.request_id)
+
+def get_user_info():
+    headers = st.context.headers
+    return dict(
+        user_name=headers.get("X-Forwarded-Preferred-Username"),
+        user_email=headers.get("X-Forwarded-Email"),
+        user_id=headers.get("X-Forwarded-User"),
+    )
+
+user_info = get_user_info()
+
+# --- Init state ---
+if "history" not in st.session_state:
+    st.session_state.history = []
+
+st.title("🧱 Chatbot App")
+st.write("A basic chatbot using your own serving endpoint.")
+
+def render_assistant_message_feedback(i, request_id):
+    def save_feedback(index):
+        submit_feedback(
+            endpoint=SERVING_ENDPOINT,
+            request_id=request_id,
+            rating=st.session_state[f"feedback_{index}"]
+        )
+    selection = st.feedback("thumbs", key=f"feedback_{i}", on_change=save_feedback, args=[i])
+
+
+# --- Render chat history ---
+for i, element in enumerate(st.session_state.history):
+    element.render(i)
+
+# --- Chat input (must run BEFORE rendering messages) ---
+if prompt := st.chat_input("Ask a question"):
+    # Add user message to chat history
+    user_msg = UserMessage(content=prompt)
+    st.session_state.history.append(user_msg)
```

Comment:
can we check if `session_state` history is stored on browser refresh. If it is we should probably have a way to reset chat. If not we can probably just have a local variable for this

Reply:
https://docs.streamlit.io/develop/api-reference/caching-and-state/st.session_state helped me understand what's going on - session state is actually used to preserve conversation history across re-renders of the app within the same page load